### PR TITLE
WIP Shell probes no longer use temp only files (PRELUDE_DIR)

### PR DESCRIPTION
**Pull request by privateducky**

### CODE REVIEW

**Comment by privateducky** [comment]
@makk94 can you help align raindrop to this change set - so it matches exactly?

File: shell/probe/nocturnal.sh [hunk 1]
```diff
@@ -1,30 +1,29 @@
...
-            if test -f "$temp";then
-                $temp
-                dat="${test}:$?"
-                $temp -cleanup
-            else
-                dat="${test}:9"
+        if [ -z "$PRELUDE_CA" ] || [ "$PRELUDE_CA" == "$ca" ];then
+            if test -f $exe;then
+                chmod +x $exe
+                $exe
```
Comment (mireaulf) [inline shell/probe/nocturnal.sh:21]: we don't run cleanup?

Comment (privateducky) [reply shell/probe/nocturnal.sh:21]: I'm toying with this now; leaning toward using 'defer' inside our tests instead, to do cleanup after test executions. why? the 'my test was quarantined during execution' edge case makes it a toss up in my mind but defer seems more likely to work in most cases.

Comment (mireaulf) [reply shell/probe/nocturnal.sh:21]: Just wanted to make sure you didn't forget about cleanup :)

**Comment by makk94** [comment]
> @makk94 can you help align raindrop to this change set - so it matches exactly?

don't we need the suffix on windows tests to be `.exe` ?

File: shell/probe/nocturnal.sh [hunk 2]
```diff
@@ -1,30 +1,29 @@
...
-    test=$(echo $location | grep -o '[0-9a-f]\{8\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{12\}' | head -n 1)
-
-    if [ -z "$test" ];then
-        sleep 14400
-    else
+do  
+    exe=$PRELUDE_DIR/$(uuidgen).vst
+    location=$(curl -sL -w %{url_effective} -o $exe -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $PRELUDE_API)
+    
+    if [ -s $exe ];then
```
Comment (makk94) [inline shell/probe/nocturnal.sh:14]: existence of the file here doesn't mean it's a test in there

ex. this throws any error message in the file too

```
mahinakaholokula@Mahinas-MacBook-Pro my-prelude % curl -sL -w %{url_effective} -o tryme.vst -H "token:itsabadone" -H "dos:darwin-arm64" -H "dat:" "https://api.preludesecurity.com"
https://api.preludesecurity.com/%  
mahinakaholokula@Mahinas-MacBook-Pro my-prelude % cat tryme.vst 
{"Message":"User is not authorized to access this resource"}%            
```

File: shell/probe/nocturnal.sh [hunk 3]
```diff
@@ -1,30 +1,29 @@
...
-        if [ -z "$PRELUDE_CA" ] || [ "$PRELUDE_CA" == "$ca" ];then
-            chmod +x $temp
 
-            if test -f "$temp";then
-                $temp
-                dat="${test}:$?"
-                $temp -cleanup
-            else
-                dat="${test}:9"
+        if [ -z "$PRELUDE_CA" ] || [ "$PRELUDE_CA" == "$ca" ];then
```
Comment (makk94) [inline shell/probe/nocturnal.sh:18]: the probe will also spin forever (skipping sleep) if this CA check fails

File: shell/probe/nocturnal.sh [hunk 4]
```diff
@@ -1,30 +1,29 @@
...
+            if test -f $exe;then
```
Comment (makk94) [inline shell/probe/nocturnal.sh:19]: should the dat assignment be in this `if` too?

Comment (makk94) [reply shell/probe/nocturnal.sh:19]: is this `if` even necessary? You already check the file exists earlier with the `[ -s $exe ] `. Seems like this check here can be combined with that one

File: shell/probe/nocturnal.sh (hunk 2 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh]: good point. sticking with $test is probably best. reverted.

File: shell/probe/nocturnal.sh (hunk 3 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:18]: ah, because it'll re-request. 
that exists today but I'll see if I can revise that in my change set here.

Comment (makk94) [reply shell/probe/nocturnal.sh:18]: might I recommend something like this in the while
```
if (test) 
    if (ca passes)
         yay run test
         continue
     fi
fi
clean up files
sleep
```

File: shell/probe/nocturnal.sh (hunk 4 above)
Comment (makk94) [reply shell/probe/nocturnal.sh:21]: i see, guess it's possible it was removed in the very small amount of time between downloading and running.

And the fallback value is always 0 because that comes from the `test -f ` call, right?

File: shell/probe/nocturnal.sh [hunk 5]
```diff
@@ -1,30 +1,32 @@
...
+            
+            if test -f $exe;then
+                echo "[P] Running $test masquerading as $exe"
+                $exe
             fi
+            dat="${test}:$?"
         fi
+    else
+        echo "[P] Cleaning up temporary cache"
+        find $PRELUDE_DIR -type f -name "*" -mmin -1 -delete
```
Comment (makk94) [inline shell/probe/nocturnal.sh:29]: don't really need the `-name *` anymore

this doesn't seem particularly scalable (our tests timeout at 10s so if someones setup sucks, they might not finish all tests in 1min. Especially as we keep adding more tests)

also a little scary to "delete all", esp. for a directory as vaguely named as "cache"

File: shell/probe/nocturnal.sh [hunk 6]
```diff
@@ -1,30 +1,32 @@
...
-            if test -f "$temp";then
-                $temp
-                dat="${test}:$?"
-                $temp -cleanup
-            else
-                dat="${test}:9"
+        if [ -z "$PRELUDE_CA" ] || [ "$PRELUDE_CA" == "$ca" ];then
+            chmod +x $exe
+            
+            if test -f $exe;then
```
Comment (makk94) [inline shell/probe/nocturnal.sh:21]: nit: use brackets like everywhere else

File: shell/probe/nocturnal.sh [hunk 7]
```diff
@@ -1,30 +1,32 @@
...
+dos=$(echo $(uname -s)-$(uname -m))
 
 while :
-do
-    temp=$(mktemp)
-    location=$(curl -sL -w %{url_effective} -o $temp -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $PRELUDE_API)
-    test=$(echo $location | grep -o '[0-9a-f]\{8\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{12\}' | head -n 1)
+do  
+    exe=$PRELUDE_DIR/$(uuidgen)
+    location=$(curl -sL -w %{url_effective} --create-dirs -o $exe -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $PRELUDE_API)
```
Comment (makk94) [inline shell/probe/nocturnal.sh:12]: am i crazy or do we need to reset dat to empty string when we sleep?

I don't see this in our code today, so hoping I'm crazy...

File: shell/probe/nocturnal.sh [hunk 8]
```diff
@@ -1,30 +1,32 @@
 #!/bin/bash
 
 PRELUDE_API=${PRELUDE_API:="https://api.preludesecurity.com"}
+PRELUDE_DIR=${PRELUDE_DIR:="./cache"}
+PRELUDE_SLEEP=${PRELUDE_SLEEP:=14440}
 
-sys=$(uname -s)-$(uname -m)
-dos=$(echo $sys | tr '[:upper:]' '[:lower:]')
+dos=$(echo $(uname -s)-$(uname -m))
```
Comment (makk94) [inline shell/probe/nocturnal.sh:7]: dont need the `echo` bit

File: shell/probe/nocturnal.sh [hunk 9]
```diff
@@ -1,30 +1,33 @@
 #!/bin/bash
 
+PRELUDE_SLEEP=${PRELUDE_SLEEP:=14440}
+PRELUDE_DIR=${PRELUDE_DIR:="./cache"}
 PRELUDE_API=${PRELUDE_API:="https://api.preludesecurity.com"}
+PRELUDE_CA=${PRELUDE_CA:="prelude-account-prod-us-west-2.s3.amazonaws.com"}
```
Comment (makk94) [inline shell/probe/nocturnal.sh:6]: arent we us-west-1

File: shell/probe/nocturnal.sh (hunk 5 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:29]: yeah; I've been working on this area a bit. going between a few thoughts atm.

File: shell/probe/nocturnal.sh (hunk 7 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:13]: good question.. I'll look into that (both pre/post this change)

File: shell/probe/nocturnal.sh (hunk 8 above)
Comment (makk94) [reply shell/probe/nocturnal.sh:8]: why are you keeping the echo ?

File: shell/probe/nocturnal.sh [hunk 10]
```diff
@@ -1,30 +1,32 @@
 #!/bin/bash
 
-PRELUDE_API=${PRELUDE_API:="https://api.preludesecurity.com"}
+PRELUDE_SLEEP=${PRELUDE_SLEEP:=14440}
+PRELUDE_DIR=${PRELUDE_DIR:="./cache"}
+PRELUDE_CA=${PRELUDE_CA:="prelude-account-prod-us-west-1.s3.amazonaws.com"}
 
-sys=$(uname -s)-$(uname -m)
-dos=$(echo $sys | tr '[:upper:]' '[:lower:]')
+api="https://api.preludesecurity.com"
```
Comment (makk94) [inline shell/probe/nocturnal.sh:7]: pleeease keep the ability to override for staging

File: shell/probe/nocturnal.sh [hunk 11]
```diff
@@ -1,30 +1,32 @@
...
 while :
-do
-    temp=$(mktemp)
-    location=$(curl -sL -w %{url_effective} -o $temp -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $PRELUDE_API)
-    test=$(echo $location | grep -o '[0-9a-f]\{8\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{12\}' | head -n 1)
+do  
+    exe=$PRELUDE_DIR/$(uuidgen)
+    location=$(curl -sL -w %{url_effective} --create-dirs -o $exe -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $api)
+    test=$(echo $location | grep -o '[0-9a-f]\{8\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{4\}-[0-9a-f]\{12\}' | head -n 1)    
+    unset dat
```
Comment (makk94) [inline shell/probe/nocturnal.sh:15]: this unset should be near the sleep, no?

File: shell/probe/nocturnal.sh [hunk 12]
```diff
@@ -1,30 +1,32 @@
...
-            else
-                dat="${test}:9"
-            fi
+        if [ "$PRELUDE_CA" == "$ca" ];then
+            echo "[P] Running $test masquerading as $exe"
+            chmod +x $exe
+            $exe
+            dat="${test}:$?"
+        else
+            echo "[P] Authority mismatch: $exe" && exit 0
```
Comment (makk94) [inline shell/probe/nocturnal.sh:26]: wouldn't you rather print $ca ?

File: shell/probe/nocturnal.sh (hunk 8 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:8]: are you looking at an old commit? there are only 3 echo statements currently in this PR, all debug statements that may remain.

File: shell/probe/nocturnal.sh (hunk 12 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:26]: good point; will update

File: shell/probe/nocturnal.sh (hunk 11 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:15]: could go either way; not sure I care but will examine both

File: shell/probe/nocturnal.sh (hunk 10 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:7]: downside is someone can use an env var to override where tests are coming from, so I'll likely opt not to keep that in.

File: shell/probe/nocturnal.sh (hunk 8 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:8]: good call

Comment (makk94) [reply shell/probe/nocturnal.sh]: have you maybe not committed what's on your local?

I'm seeing this
Screenshot 2023-03-07 at 4 53 31 PM

File: shell/probe/nocturnal.sh (hunk 10 above)
Comment (makk94) [reply shell/probe/nocturnal.sh:7]: isn't that what the CA check is for?

File: shell/probe/nocturnal.sh (hunk 11 above)
Comment (makk94) [reply shell/probe/nocturnal.sh]: thing I'm thinking about that needs the unset near the sleep:

iteration 1:
 - get a test, set dat={123-123-123:100}
 iteration 2:
 - send in that dat and get back "no more tests"
 - go to sleep
 iteration 3:
- exit sleep, dat still set to {123-123-123:100}
- send in that dat while getting next round of tests

File: shell/probe/nocturnal.sh (hunk 10 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:7]: sort of... two different things.. but the CA check may override the concern as you're pointing out... I think you're spot on

File: shell/probe/nocturnal.sh (hunk 9 above)
Comment (makk94) [reply shell/probe/nocturnal.sh]: you're looking at an old comment ;)

File: shell/probe/nocturnal.sh (hunk 11 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh]: on iteration 2, when you get 'no more tests', you'd be unsetting the 'data' before getting to the if/check on the $test.

File: shell/probe/nocturnal.sh [hunk 13]
```diff
@@ -1,30 +1,33 @@
 #!/bin/bash
 
+PRELUDE_SLEEP=${PRELUDE_SLEEP:=14440}
+PRELUDE_DIR=${PRELUDE_DIR:="./cache"}
```
Comment (makk94) [inline shell/probe/nocturnal.sh:4]: could we at least call this somethign more prelude-specific?

File: shell/probe/nocturnal.sh (hunk 11 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh]: (but I swapped to sleep anyway, as I think it reads better)

File: shell/probe/nocturnal.sh (hunk 13 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:4]: yea, still thinking on this new variable overall.. got a name suggestion vs cache?

File: shell/probe/nocturnal.sh [hunk 14]
```diff
@@ -1,30 +1,33 @@
...
-dos=$(echo $sys | tr '[:upper:]' '[:lower:]')
+dos=$(uname -s)-$(uname -m)
 
 while :
-do
-    temp=$(mktemp)
-    location=$(curl -sL -w %{url_effective} -o $temp -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $PRELUDE_API)
+do  
+    exe=$PRELUDE_DIR/$(uuidgen)
+    location=$(curl -sL -w %{url_effective} --create-dirs -o $exe -H "token:${PRELUDE_TOKEN}" -H "dos:${dos}" -H "dat:${dat}" $PRELUDE_API)
```
Comment (makk94) [inline shell/probe/nocturnal.sh:13]: will you make the curl flags include `curl -sfSL` so it'll print out any http error codes

File: shell/probe/nocturnal.sh (hunk 13 above)
Comment (makk94) [reply shell/probe/nocturnal.sh:4]: im bad at names ha but presumably something including the work "prelude" unless there's a reason you're avoiding it

File: shell/probe/nocturnal.sh (hunk 14 above)
Comment (privateducky) [reply shell/probe/nocturnal.sh:13]: hmm, where would those print to in that case, as we're mapping to $location?

Comment (makk94) [reply shell/probe/nocturnal.sh]: the -f has the  http failure code print to stdout. location is still filled with url_effective

the -S isn't as necessary now that we're deleting the files, but that stop the file from being created on http error code. The --create-dirs still takes effect tho

File: shell/probe/nocturnal.sh (hunk 10 above)
Comment (makk94) [reply shell/probe/nocturnal.sh:7]: why'd this go back :(

Comment (privateducky) [reply shell/probe/nocturnal.sh:7]: can't leave API and CA both mutable
